### PR TITLE
[ember] refactor @ember/object types into their own package

### PR DESCRIPTION
**Pull request by mike-north**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
> this is an odd situation where the npm package and the name of the modules consumers import do not align. These types align with the module names

- Create it with `dts-gen --dt`, not by basing it on an existing project.
> Does not apply in this case

- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://www.emberjs.com/api/ember/3.4/modules/@ember%2Fobject
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

cc: @dwickern @jamescdavis @dfreeman @chriskrycho 

Resolves https://github.com/typed-ember/ember-cli-typescript/issues/272

### CODE REVIEW

File: types/ember__object/test/access-modifier.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    protected bar() { return 'bar'; }
+    private baz() { return 'baz'; }
+}
+const f = new Foo();
+assertType<string>(f.hello());
+assertType<string>(f.bar()); // $ExpectError
+assertType<string>(f.baz()); // $ExpectError
+
+// TODO: enable after TS 3.0 https://github.com/typed-ember/ember-cli-typescript/issues/291
+// class Foo2 extends EmberObject.extend({
```
Comment (mike-north) [inline types/ember__object/test/access-modifier.ts]: Enable this after https://github.com/typed-ember/ember-cli-typescript/issues/291 is resolved

File: types/ember__object/test/core.ts [hunk 2]
```diff
@@ -0,0 +1,198 @@
+import Ember from 'ember';
+import { assertType } from './lib/assert';
+
+/** Newable tests */
+const co1 = new Ember.CoreObject();
+
+// TODO: Enable in TS 3.0  see: https://github.com/typed-ember/ember-cli-typescript/issues/291
```
Comment (mike-north) [inline types/ember__object/test/core.ts]: Enable this after https://github.com/typed-ember/ember-cli-typescript/issues/291 is resolved

File: types/ember__object/test/core.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+// co1.concatenatedProperties; // $ExpectType string[]
+co1.isDestroyed; // $ExpectType boolean
+co1.isDestroying; // $ExpectType boolean
+co1.destroy(); // $ExpectType CoreObject
+co1.toString(); // $ExpectType string
+
+/** .create tests */
+const co2 = Ember.CoreObject.create();
+// TODO: Enable in TS 3.0  see: https://github.com/typed-ember/ember-cli-typescript/issues/291
```
Comment (mike-north) [inline types/ember__object/test/core.ts]: Enable this after https://github.com/typed-ember/ember-cli-typescript/issues/291 is resolved

File: types/ember__object/test/core.ts [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        this.bee; // $ExpectError
+    }
+}, {
+    foo: 99,
+    bee: 'honey',
+    func2() {
+        // this includes stuff from CoreObject
+        this.init; // $ExpectType () => void
+        // this includes stuff from this extend-arg
+        // TODO: switch to "$ExpectType number" in TS 3.0  see: https://github.com/typed-ember/ember-cli-typescript/issues/291
```
Comment (mike-north) [inline types/ember__object/test/core.ts]: Enable this after https://github.com/typed-ember/ember-cli-typescript/issues/291 is resolved

File: types/ember__object/test/core.ts [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        this.foo; // $ExpectType string & number
+        // this includes stuff from earlier extend-args
+        this.bar; // $ExpectType number
+    }
+}).create();
+
+// TODO: enable in TS 3.0  see: https://github.com/typed-ember/ember-cli-typescript/issues/291
```
Comment (mike-north) [inline types/ember__object/test/core.ts]: Enable this after https://github.com/typed-ember/ember-cli-typescript/issues/291 is resolved

File: types/ember__object/test/core.ts [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        // this includes stuff from CoreObject
+        this.init; // $ExpectType () => void
+        // this includes stuff from this extend-arg
+        this.money; // $ExpectType string
+        // this includes stuff from earlier extend-args
+        this.bee; // $ExpectType string
+        this.bar; // $ExpectType number
+    }
+}).create();
+// TODO: enable in TS 3.0  see: https://github.com/typed-ember/ember-cli-typescript/issues/291
```
Comment (mike-north) [inline types/ember__object/test/core.ts]: Enable this after https://github.com/typed-ember/ember-cli-typescript/issues/291 is resolved

File: types/ember__object/test/core.ts [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        this.bee; // $ExpectError
+    }
+}, {
+    foo: 99,
+    bee: 'honey',
+    func2() {
+        // this includes stuff from CoreObject
+        this.init; // $ExpectType () => void
+        // this includes stuff from this extend-arg
+        // TODO: switch to "$ExpectType number" in TS 3.0  see: https://github.com/typed-ember/ember-cli-typescript/issues/291
```
Comment (mike-north) [inline types/ember__object/test/core.ts]: Enable this after https://github.com/typed-ember/ember-cli-typescript/issues/291 is resolved

File: types/ember__object/test/core.ts [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        // this includes stuff from this extend-arg
+        this.neighborhood; // $ExpectType string
+        // this includes stuff from earlier extend-args
+        this.bee; // $ExpectType string
+        this.bar; // $ExpectType number
+        this.money; // $ExpectType string
+    }
+}).create();
+
+// TODO: enable in TS 3.0  see: https://github.com/typed-ember/ember-cli-typescript/issues/291
```
Comment (mike-north) [inline types/ember__object/test/core.ts]: Enable this after https://github.com/typed-ember/ember-cli-typescript/issues/291 is resolved

File: types/ember__object/test/event.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    }).one('greet', () => {
+        console.log('Offer one-time special');
+    }).off('event', {}, () => {});
+
+    person.greet();
+}
+
+function testObserver() {
+    EmberObject.extend({
+        // TODO: enable after https://github.com/typed-ember/ember-cli-typescript/issues/290
```
Comment (mike-north) [inline types/ember__object/test/event.ts]: Enable this after https://github.com/typed-ember/ember-cli-typescript/issues/290 is resolved

File: types/ember__object/test/object.ts [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        this.setProperties({ b: 11 });
+        // this.setProperties({ b: '11' }); // $ExpectError
+        this.setProperties({
+            a: 'def',
+            b: 11
+        });
+    }
+}
+
+// TODO: enable after TS 3.0 https://github.com/typed-ember/ember-cli-typescript/issues/291
```
Comment (mike-north) [inline types/ember__object/test/object.ts]: Enable this after https://github.com/typed-ember/ember-cli-typescript/issues/291 is resolved

**Comment by RyanCavanaugh** [comment]
@mike-north good to merge?

**Comment by mike-north** [comment]
@RyanCavanaugh yes, good to merge

**Pull request merged**